### PR TITLE
Clustering he

### PR DESCRIPTION
**Pull request by zhut19**
Merging my fork to xenon1t/pax clustering HE branch.
This merge include changes done in pulse desaturation.

### CODE REVIEW

File: pax/plugins/signal_processing/DesaturatePulses.py [hunk 1]
```diff
@@ -27,71 +27,117 @@ def transform_event(self, event):
...
-            reference_slice = slice(max(0, first_saturated - self.config['reference_region_samples']),
-                                    first_saturated)
             # Find all pulses in TPC channels that overlap with the saturated & reference region
             other_pulses = [p for i, p in enumerate(event.pulses)
-                            if p.left < last_saturated + pulse.left and p.right > pulse.left and
+                            if p.left < pulse.right and p.right > pulse.left and
                             not is_saturated[i] and
-                            p.channel in tpc_channels]
+                            p.channel in tpc_channels and
+                            p.channel not in self.config['large_after_pulsing_channels']]
```
Comment (feigaodm) [inline pax/plugins/signal_processing/DesaturatePulses.py]: What if only those large after pulsing channels are not saturated? I guess it's not very likely?

File: pax/plugins/signal_processing/DesaturatePulses.py [hunk 2]
```diff
@@ -27,71 +27,117 @@ def transform_event(self, event):
...
+                if len(sumw[reference_slice][sumw[reference_slice] > 1]) \
+                        < self.config['reference_region_samples_treshold']:
+                    # the pulse is saturated, but there are not enough reference samples to get a good ratio
+                    # This actually distinguished between S1 and S2 and will only correct S2 signals
+                    continue
+
+                ratio = w[reference_slice].sum()/sumw[reference_slice].sum()
+
+                # not < is preferred over >, since it will catch nan
+                if not ratio < self.config.get('min_reference_area_ratio', 1):
```
Comment (feigaodm) [inline pax/plugins/signal_processing/DesaturatePulses.py]: would be good to add a link to a plot for each exception, and put them in your note.

File: pax/plugins/signal_processing/DesaturatePulses.py (hunk 1 above)
Comment (zhut19) [reply pax/plugins/signal_processing/DesaturatePulses.py]: Doesn't seem likely, but not sure what will happen if other_pulses happen to be empty.

**Comment by sanderbreur** [comment]
Hello, I don't know what this improvement will do, but would love to see some waveform plot's to see the improvement :)

**Comment by feigaodm** [comment]
@sanderbreur two notes (from @zhut19 and @chiaracapelli ) are prepared for a discussion tomorrow, welcome to join the signal reconstruction group :)

**Comment by sanderbreur** [comment]
Cool, perfect

**Pull request merged**